### PR TITLE
Feat/optimize new communities

### PR DESCRIPTION
**Pull request by brenzi**
@pifragile please split this PR into smaller single-topic ones: 
* setup benchmarking for communities pallet
* change O(n^2) location distance checks algo to ~O(n) geohash bucket approach

### CODE REVIEW

File: communities/src/lib.rs [hunk 1]
```diff
@@ -92,15 +113,31 @@ decl_module! {
                 validate_nominal_income(&i).map_err(|_| <Error<T>>::InvalidNominalIncome)?;
             }
 
-            let cid = CommunityIdentifier::from(blake2_256(&(loc.clone(), bootstrappers.clone()).encode()));
+            let cid = CommunityIdentifier::from(blake2_256(&(location.clone(), bootstrappers.clone()).encode()));
```
Comment (brenzi) [inline communities/src/lib.rs:116]: I wonder if it makes sense to change the derivation for cid.
Would it make sense to let the first few bytes of the cid type by the geohash? This way we could use the cid value in various ways as it contains information and the bucket could be derived directly. The GeoHash type could directly offer a `fn get_bucket()` and we could drop the storage map `CommunityIdentifiersByGeohash`. less state bloat.
thoughts? (also @clangenb )

I think the geohash should be more precise than `GEO_HASH_LENGTH` used for buckets (AFAIU) because that would make it even easier to diplay pins on a map in encointer/explorer without looking up locations at all

I would still postfix a hash of (location, bootstrappers) to bind these informations

Comment (brenzi) [reply communities/src/lib.rs:116]: nice thing: from the cid alone in ss58 representation you could learn to guess the geographical region of the community

File: communities/src/lib.rs [hunk 2]
```diff
@@ -92,15 +113,31 @@ decl_module! {
...
+            let geo_hash = encode(location.lat, location.lon, GEO_HASH_LENGTH).map_err(|_| <Error<T>>::InvalidLocationForGeohash)?;
+            let mut locations: Vec<Location> = Vec::new();
+
+            // insert cid into cids_by_geohash map
+            let mut cids_in_bucket = Self::cids_by_geohash(geo_hash.to_pallet_string());
+            match cids_in_bucket.binary_search(&cid) {
+                Ok(_) => (),
+                Err(index) => {
+                    cids_in_bucket.insert(index, cid.clone());
+                    <CommunityIdentifiersByGeohash>::insert(geo_hash.to_pallet_string(), cids_in_bucket);
```
[the remark this answers is not in the document]
Comment (brenzi) [reply communities/src/lib.rs:131]: everywhere....

File: communities/src/lib.rs (hunk 1 above)
Comment (brenzi) [reply communities/src/lib.rs:116]: and the `Locations`storage could probably become a `map` instead of `double_map`, right?

Comment (brenzi) [reply communities/src/lib.rs:116]: as this would be a fundamental refactoring, I believe we need to decide this first....as the rest of the review will depend on it

File: communities/src/lib.rs [hunk 3]
```diff
@@ -92,15 +113,31 @@ decl_module! {
...
 
-            let cid = CommunityIdentifier::from(blake2_256(&(loc.clone(), bootstrappers.clone()).encode()));
+            let cid = CommunityIdentifier::from(blake2_256(&(location.clone(), bootstrappers.clone()).encode()));
             let cids = Self::community_identifiers();
             ensure!(!cids.contains(&cid), "community already registered");
-            Self::validate_locations(&loc, &cids)?; // <- O(n^2) !!!
 
+            Self::validate_location(&location)?;
             // All checks done, now mutate state
+            let geo_hash = encode(location.lat, location.lon, GEO_HASH_LENGTH).map_err(|_| <Error<T>>::InvalidLocationForGeohash)?;
```
[the remark this answers is not in the document]
Comment (brenzi) [reply communities/src/lib.rs:122]: GEO_HASH_LENGTH should rather be called BUCKET_RESOLUTION I think

File: communities/src/lib.rs [hunk 4]
```diff
@@ -92,15 +113,31 @@ decl_module! {
...
+            let mut locations: Vec<Location> = Vec::new();
+
+            // insert cid into cids_by_geohash map
+            let mut cids_in_bucket = Self::cids_by_geohash(geo_hash.to_pallet_string());
+            match cids_in_bucket.binary_search(&cid) {
```
Comment (brenzi) [inline communities/src/lib.rs:127]: is this better than `.contains()`? more performant?

Comment (brenzi) [reply communities/src/lib.rs:127]: I'd rather use `if` than `match`

File: communities/src/lib.rs [hunk 5]
```diff
@@ -114,6 +151,60 @@ decl_module! {
             info!(target: LOG, "registered community with cid: {:?}", cid);
         }
 
+        #[weight = 10_000]
+        pub fn add_location(origin, cid: CommunityIdentifier, location: Location) {
```
Comment (brenzi) [inline communities/src/lib.rs:155]: I believe we should support adding several locations with one extrinsic. But we'll have to enforce a maximum number

File: communities/src/lib.rs [hunk 6]
```diff
@@ -264,50 +357,110 @@ impl<T: Config> Module<T> {
...
-                for l2 in Self::locations(other) {
-                    if Self::solar_trip_time(&l1, &l2) < MIN_SOLAR_TRIP_TIME_S {
-                        warn!(target: LOG,
-                                     "location {:?} too close to previously registered location {:?} with cid {:?}",
-                                     l1, l2, other);
-                        return Err(<Error<T>>::MinimumDistanceViolationToOtherCommunity)?;
-                    }
-                }
+
+        //check if northern neighbour bucket needs to be included
```
Comment (brenzi) [inline communities/src/lib.rs:370]: this block contains massively redundant code. But I guess the only way to reduce this would be a macro. Might not be worth bothering....

File: communities/src/lib.rs (hunk 4 above)
Comment (pifragile) [reply communities/src/lib.rs:127]: i used what substrate proposed [here](https://substrate.dev/recipes/vec-set.html)

File: communities/src/benchmarking.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+benchmarks! {
+    new_community {
+    let i in 2 .. NUM_LOCATIONS;
+    warn!("starting benchmark {:?}", i);
+    let caller: T::AccountId = whitelisted_caller();
+
+    let bootstrappers : Vec<T::AccountId> = (0..12).map(|n| account("dummy name", n, n)).collect();
+    let mut community_metadata = CommunityMetadata::default();
+    community_metadata.name = "20charsaaaaaaaaaaaaa".into();
+    community_metadata.url = Some("19charsaaaaaaaaa.ch".into());
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/benchmarking.rs:37]: the current validation code enforces a maximum of 19 chars.
https://github.com/encointer/pallets/blob/cfd318ef91d18059ad61881fe2bdae69e42252ae/primitives/src/communities.rs#L132

File: communities/src/lib.rs [hunk 8]
```diff
@@ -70,13 +94,10 @@ decl_module! {
...
         fn deposit_event() = default;
         type Error = Error<T>;
-        // FIXME: this function has complexity O(n^2)!
-        // where n is the number of all locations of all communities
-        // this should be run off-chain in substraTEE-worker later
         #[weight = 10_000]
         pub fn new_community(
             origin,
-            loc: Vec<Location>,
+            location: Location,
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs:100]: exactly

File: communities/Cargo.toml [hunk 9]
```diff
@@ -62,6 +69,11 @@ version = "1.0.101"
 [dev-dependencies]
 approx = "0.3.0"
 
+[dependencies.geohash]
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/Cargo.toml:72]: done

File: communities/src/benchmarking.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (clangenb) [reply communities/src/benchmarking.rs:36]: There might have been some misunderstanding from my side.

Currently, the Icons (CommunityTokenThumbnails) are stored in ipfs, whereas the url is a regular link to a website representing the community.

However, I realize now that 19 chars is probably to few characters anyhow.

File: communities/src/lib.rs (hunk 1 above)
Comment (pifragile) [reply communities/src/lib.rs:116]: @brenzi  i am not sure if this is feasible, because in order to find all locations in one bucket, we would have to loop over all cids and do a prefix matching, which is pretty inefficient.

about the geo hash precision: it might be better for visual purposes to increase the geohash precision, but with increasing precision, we would have to do significantly more storage lookups in order to find buckets in a certain radius, which would also make things less efficient again.

File: communities/src/lib.rs [hunk 10]
```diff
@@ -31,10 +31,13 @@ use frame_support::{
 };
 use frame_system::{ensure_root, ensure_signed};
 use rstd::prelude::*;
+use rstd::result::Result;
 use fixed::transcendental::{asin, cos, powi, sin, sqrt};
 use runtime_io::hashing::blake2_256;
 use sp_runtime::{DispatchResult, SaturatedConversion};
 
+use geohash::{neighbors, encode, decode};
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs:39]: i do not fully understand. should the new `GeoHash` type just be a string type implementing the above mentioned `Encode` and `Decode` or should it include also the full functionality? @clangenb can you maybe help me on this?

File: communities/src/lib.rs [hunk 11]
```diff
@@ -45,6 +48,22 @@ use encointer_primitives::{
...
+    fn to_pallet_string(&self) -> PalletString {
+        self.clone()
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn to_pallet_string(&self) -> PalletString {
+        return (self.clone().into_bytes()).to_vec();
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs]: done

File: communities/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs]: done

File: communities/src/lib.rs [hunk 12]
```diff
@@ -54,7 +73,12 @@ const LOG: &str = "encointer";
 
 decl_storage! {
     trait Store for Module<T: Config> as EncointerCommunities {
-        Locations get(fn locations): map hasher(blake2_128_concat) CommunityIdentifier => Vec<Location>;
+        // map geohash -> cid
+        CommunityIdentifiersByGeohash get(fn cids_by_geohash): map hasher(identity) PalletString => Vec<CommunityIdentifier>;
+
+        // map cid -> geohash -> location
+        Locations get(fn locations): double_map hasher(blake2_128_concat) CommunityIdentifier, hasher(identity) PalletString => Vec<Location>;
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs]: done

File: communities/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs]: done after conversation with @clangenb

File: communities/src/benchmarking.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/benchmarking.rs:36]: will resolve this as it seems  only loosely related to this PR

File: communities/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/lib.rs]: done, side note `encode` was the actual geohash encoding (from coordinates to geohash). this confusion is now also resolved

File: communities/src/lib.rs (hunk 1 above)
Comment (pifragile) [reply communities/src/lib.rs:100]: @brenzi and there is also another flaw in this approach: one community will most of the time spread over more than one geohash bucket

File: ceremonies/src/lib.rs [hunk 13]
```diff
@@ -373,7 +373,9 @@ impl<T: Config> Module<T> {
         for cid in cids.iter() {
             let pcount = <ParticipantCount>::get((cid, cindex));
             let ecount = <EndorseesCount>::get((cid, cindex));
-            let n_locations = <encointer_communities::Module<T>>::locations(cid).len();
+            // FIXME
```
[the remark this answers is not in the document]
Comment (pifragile) [reply ceremonies/src/lib.rs:376]: yep, we will have to talk about this in more detail, because location assignment and issuance depend on the `communities` pallet and therefore our design changes here will have implications.

File: communities/src/benchmarking.rs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    balances:: {Demurrage, consts::DEFAULT_DEMURRAGE}
+};
+
+use log::{info, warn};
+
+const NUM_LOCATIONS : u32 =  500000;
+
+fn get_location(i:u32) -> Location {
+    // splits the world into 1m locations
+    let max_locations = 1000000;
```
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/benchmarking.rs]: true, the locations are not really representative. when working on the benchmarks i was focussing more on comparing the two algorithms than on finding the actual worst case weights for the extrinsic. but i agree that we could optimize the benchmarks in order to find the weights.

we could change the benchmarks to only generate a limited set of locations within one bucket and all of its neighbor buckets and we can drop all the other locations, when we are not comparing with the `O(n)` algorithm. Like this we can really test for all the edge cases of the algorithm in order to find the worst case weights.

i propose that we do this in a separate PR.

File: communities/src/lib.rs (hunk 1 above)
Comment (clangenb) [reply communities/src/lib.rs:100]: I agree so far with @pifragile. I do not see a better alternative to the current storage design. However, I like the direct incorporation of the GeoHash in the cid, which we could do even if we can't use this information.

File: communities/src/lib.rs (hunk 4 above)
Comment (clangenb) [reply communities/src/lib.rs:111]: Although verbose, I reckon this is indeed more efficient.

File: communities/src/lib.rs [hunk 15]
```diff
@@ -264,50 +341,110 @@ impl<T: Config> Module<T> {
...
+        // solar time = time human - time sun
+        // width of one bucket at latitude x = 4900 * cos(x) m
+        // speed human: 83 m/s
+        // speed sun: (cos(x) * 111319) / 240 = (meters/degree)/(seconds/degree) = m/s
+        // time sun = (4900 * cos(x)) / ((cos(x) * 111319) / 240)
+        // time human = (4900 * cos(x)) / 83
+        // solar time = ((4900 * cos(x)) / 83) - ((4900 * cos(x)) / ((cos(x) * 111319) / 240))
+        // solve ((4900 * cos(x)) / 83) - ((4900 * cos(x)) / ((cos(x) * 111319) / 240)) = 1
+        // x = 78.7036, so below 78.7036 it takes a human more than 1 second solar time to
+        // traverse 1 bucket horizontally
```
Comment (clangenb) [inline communities/src/lib.rs:378]: I like the explanation. 🚀

**Comment by pifragile** [comment]
> @pifragile please split this PR into smaller single-topic ones:
> 
> * setup benchmarking for communities pallet
> * change O(n^2) location distance checks algo to ~O(n) geohash bucket approach

@brenzi this will be a little tough as the second PR will depend on the first!
what is the purpose of the split? do you just want to only merge one of them?
do you want to merge them progressively? then i could open up the first, and after it is merged, i open the second.

or do you want to educate me not to open multi-topic PRs? in that case i will remember it for the future 😉

File: communities/src/lib.rs (hunk 1 above)
Comment (clangenb) [reply communities/src/lib.rs:100]: We could argue about the hash precision, but I believe this should be solved client-side.

**Comment by clangenb** [comment]
> 
> > @pifragile please split this PR into smaller single-topic ones:
> > 
> > * setup benchmarking for communities pallet
> > * change O(n^2) location distance checks algo to ~O(n) geohash bucket approach
> 
> @brenzi this will be a little tough as the second PR will depend on the first!
> what is the purpose of the split? do you just want to only merge one of them?
> do you want to merge them progressively? then i could open up the first, and after it is merged, i open the second.
> 
> or do you want to educate me not to open multi-topic PRs? in that case i will remember it for the future 😉

I think this is also to be able to do a before-/after benchmark.

In general, we try to keep PRs as small and single-topicked as possible, as it is easier to review. It is also easier to pinpoint regression bugs.

**Comment by pifragile** [comment]
> I think this is also to be able to do a before-/after benchmark.

In order to to a benchmark at a specific revision, you can just checkout that revision and run the benchmark. Or am i understanding you wrong? btw i have already done before-/after benchmarks :) Also the benchmarks that i did had to be done on separate branches anyway, because they skip validation on benchmarking setup in order for the setup to run faster(for the old algorithm it would have taken hours if not days to setup 500000 locations) and this code should definitely not be included in `master`.

new_community_execution_time
new_community_execution_time_new


> In general, we try to keep PRs as small and single-topicked as possible, as it is easier to review. It is also easier to pinpoint regression bugs.

That makes absolute sense :) my bad.

**Comment by clangenb** [comment]
Boah, love the benchmarks!

We tend to squash merge, so we can't check out that specific revision afterwards. However, as it is not possible anyhow according to you, it does not matter.

**Comment by pifragile** [comment]
> We tend to squash merge, so we can't check out that specific revision afterwards. However, as it is not possible anyhow according to you, it does not matter.

oh right, i see!

File: ceremonies/src/lib.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (brenzi) [reply ceremonies/src/lib.rs:376]: sure. the PR may not break anything, as we only ,erge what passes CI tests

File: communities/src/benchmarking.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (brenzi) [reply communities/src/benchmarking.rs]: ok

[the remark this answers is not in the document]
Comment (brenzi) [reply communities/src/benchmarking.rs]: (please open an issue so we remember ;-)

File: communities/src/lib.rs (hunk 1 above)
Comment (brenzi) [reply communities/src/lib.rs:100]: @pifragile I think this is a misunderstanding. I first and foremost just propose to change the cid type. Nothing changes with buckets. Before, it was a meaningless hash, now I suggest to let it start with a geohash. In the minimum this gives us advantages on the UI performance side. This change should not affect anything in your logic. But it might have advantages that I'm not yet clear about, it's more a gut feeling.
I see that `get_bucket` will only lead a single bucket result, not sure if that will work

File: communities/src/benchmarking.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (pifragile) [reply communities/src/benchmarking.rs]: ok done

**Comment by brenzi** [comment]
@pifragile Yes, if you split your PR, then the second will depend on the first. But we do that all the time. As @clangenb  explained, we maintain a linear history on master and we'd need benchmarking as a separate commit. Even if you did before/after tests, we always want to be able to re-run them for whatever reason

File: communities/src/lib.rs (hunk 1 above)
Comment (pifragile) [reply communities/src/lib.rs:100]: @brenzi, oh i see. i thought you wanted to change the datastructures as well!
is this something you want to consider implementing as part of this PR, or should we leave this as food for thought?

**Comment by brenzi** [comment]
when you create the second split PR, you can just target the PR at your first PR's branch instead of master. This way, we can clearly review the changes before the first PR is merged indeed. Once the first is merged, we can retarget the second to master and rebase....and so on.

File: communities/src/lib.rs (hunk 5 above)
Comment (pifragile) [reply communities/src/lib.rs:139]: Opened issue https://github.com/encointer/pallets/issues/53 for this

File: communities/src/lib.rs (hunk 1 above)
Comment (pifragile) [reply communities/src/lib.rs:100]: Opened issue https://github.com/encointer/pallets/issues/54 for this.

**Comment by pifragile** [comment]
closed in favour of https://github.com/encointer/pallets/pull/55

**Comment by pifragile** [comment]
closed in favour of https://github.com/encointer/pallets/pull/55
